analyzers/src/SonarAnalyzer.CSharp/Rules/RedundancyInConstructorDestructorDeclaration.cs

```diff
@@ -34,6 +34,27 @@ protected override void Initialize(SonarAnalysisContext context)
         {
             context.RegisterNodeAction(CheckConstructorDeclaration, SyntaxKind.ConstructorDeclaration);
             context.RegisterNodeAction(CheckDestructorDeclaration, SyntaxKind.DestructorDeclaration);
+            context.RegisterNodeAction(CheckPrimaryConstructorRef, SyntaxKind.ClassDeclaration, SyntaxKindEx.RecordClassDeclaration);
+            context.RegisterNodeAction(CheckPrimaryConstructorVal, SyntaxKind.StructDeclaration, SyntaxKindEx.RecordStructDeclaration);
+        }
+
+        private static void CheckPrimaryConstructorRef(SonarSyntaxNodeReportingContext context) =>
+            CheckPrimaryConstructor(context, checkInitializedFieldOrProperty: false);
+
+        private static void CheckPrimaryConstructorVal(SonarSyntaxNodeReportingContext context) =>
+            CheckPrimaryConstructor(context, checkInitializedFieldOrProperty: true);
+
+        private static void CheckPrimaryConstructor(SonarSyntaxNodeReportingContext context, bool checkInitializedFieldOrProperty)
+        {
+            var typeDeclaration = (TypeDeclarationSyntax)context.Node;
+            if (TypeDeclarationSyntaxExtensions.ParameterList(typeDeclaration) is { Parameters.Count: 0 } parameterList)
+            {
+                if (checkInitializedFieldOrProperty && ContainsInitializedFieldOrProperty((INamedTypeSymbol)context.SemanticModel.GetDeclaredSymbol(context.Node)))
```

Comment:
Can't we check the syntax kind here, instead of passing `true` or `false` in the registration above, and run `ContainsInitializedFieldOrProperty` conditionally to `checkInitializedFieldOrProperty`?

And also invert the condition, to avoid the early return.

```
If SyntaxKind.ClassDeclaration or SyntaxKindEx.RecordClassDeclaration
   or !ContainsInitializedFieldOrProperty 
then context.ReportIssue
```

Reply:
I thought that instead of registering for everything and then sorting, it would be better to have a boolean flag and pass it to a method to keep that information without doing more calculations. Regarding the condition, I inverted it on purpose to simplify the reading, but if it's not the case than I can revert to the solution you suggested.

Reply:
What about the following:
```cs
private readonly CSharpFacade language = CSharpFacade.Instance;

        protected override void Initialize(SonarAnalysisContext context)
        {
            context.RegisterNodeAction(CheckConstructorDeclaration, SyntaxKind.ConstructorDeclaration);
            context.RegisterNodeAction(CheckDestructorDeclaration, SyntaxKind.DestructorDeclaration);
            context.RegisterNodeAction(CheckPrimaryConstructor, language.SyntaxKind.TypeWithPrimaryConstructorDeclarations);
        }

        private void CheckPrimaryConstructor(SonarSyntaxNodeReportingContext context)
        {
            var typeDeclaration = (TypeDeclarationSyntax)context.Node;
            if (typeDeclaration.ParameterList() is { Parameters.Count: 0 } parameterList
                && (typeDeclaration.IsAnyKind(language.SyntaxKind.ClassAndRecordClassDeclarations)
                    || !ContainsInitializedFieldOrProperty(context.SemanticModel.GetDeclaredSymbol(typeDeclaration))))
            {
                context.ReportIssue(Diagnostic.Create(Rule, parameterList.GetLocation(), "primary constructor"));
            }
        }
```

One level of nesting less, one flag less, and simpler registration to the context.
The first condition evaluated by the if (`typeDeclaration.ParameterList() is { Parameters.Count: 0 } parameterList`) is so narrow (i.e. only type declarations with primary constructors empty) that doing a little bit of work more in `typeDeclaration.IsAnyKind(language.SyntaxKind.ClassAndRecordClassDeclarations)` is not going to make any difference. Also, checking if the `SyntaxKind` of `typeDeclaration` is in an array of two elements is a very cheap operation, and since this condition is only executed on type declarations with empty parameter list, any optimization that decrease readability of the code is, in my opinion, not worth it.

I would also call the method `CheckTypeWithPrimaryConstructor`, instead of `CheckPrimaryConstructor`, because the node being registered to the callback is the type, not the primary constructor itself.

Reply:
To me, the flag with explicit registration is more readable, but since this is not a strong opinion I'll change it to what you suggested 👍